doc: add guide and answers to FAQs for first-time contributors (#63685)
nodejs/node: doc/contributing/first-contributions.md

# Guides and FAQs for first-time contributors

Welcome to the Node.js project! This guide is complementary to the
[general contribution guidelines](../../CONTRIBUTING.md)
and the [pull request guidelines](./pull-requests.md), with a focus on
clarifying the process for first-time contributors.

## Before you start

Read and understand [Code of Conduct](./code-of-conduct.md) before contributing to the project.

## Getting started with issues

If your first contribution is not motivated by an issue you already encounter, but you just want to help,
you can find issues to work on by browsing the [issue tracker](https://github.com/nodejs/node/issues).
Filter issues by the `good first issue` label to find issues that are suitable for first-time
contributors. `help wanted` tracks issues that need volunteers to help but may be more challenging
due to complexity or requiring special domain expertise.

If your first contribution is motivated by an issue you already encounter, but your proposed change
is complex, introduces a new feature, breaks backward compatibility, or needs a significant refactor,

> it's best to open an issue to discuss the change before you start working on it.

Reviewer: That contradicts https://github.com/nodejs/node/blob/040c51c98f4f3a65e36db64e2eb431294201fdf5/doc/contributing/feature-request-management.md?plain=1#L3-L8

It's also a bit confusing to use in the same sentence the word "issue" to mean two different things.

```suggestion
If your first contribution is motivated by a problem you already encounter, but your proposed change
is complex, introduces a new feature, breaks backward compatibility, or needs a significant refactor,
it's best to open find a simpler one before you start working on it. Do not open an issue for
something you're ready to tackle.
```

Author: I think that suggestion is more specifically for "features I want but I don't know how to implement", but the sentence here is more "changes I already have a prototype implementation with because I need it, but I haven't even discussed with the upstream whether this change/the design is desirable" (e.g. they may have a Node.js fork and are only trying to upstream patches; suggesting them to find a different issue is not going to helpful, because they aren't interested in tackling other irrelevant issues). In this case the issue is not a feature request, but more of a RFC/design discussion.

Reviewer: I'm not sure I understand, if they already have patches to send, wouldn't a PR be a better space to discuss it? Maybe that's a me thing, but I find reading the code much better than reading an English text – and conversely, I'm much worse at writing an issue than drafting a PR

IIUC this sentence is encouraging folks to open issues instead of PR, what would be the upside of that? Who is it for? Do we have example of succeesful contributions that started with an issue?

(FWIW for my first contribution, I started with an issue (https://github.com/nodejs/node/issues/19443) and that was a complete waste of my time)

Author: > I find reading the code much better than reading an English text

I think it depends on what change it is, hence the conditions (complex, introduces a new feature, breaks backward compatibility etc.), not just arbitrary PRs, mostly the kind of PRs where the implications are not always obvious from the code. I for one rarely jump into reviewing a complex semver-minor/major PR that does something non-trivial from a total stranger and the proposed change was never discussed elsewhere, and I think in practice these PR tends to stall for a very long time, or they get stuck in endless design discussions in the PR and abandoned after too many rounds of "do we even want to have it/do we even want it to work this way" discussion instead of "is the diff here correct to implement something we want". (see https://github.com/nodejs/node/pulls?q=is%3Apr+is%3Aopen+sort%3Acomments-desc and https://github.com/nodejs/node/pulls?q=is%3Apr+is%3Aopen+review%3Anone+sort%3Acreated-asc especially the ones from first-time contributors)

> Do we have example of succeesful contributions that started with an issue?

I think most complex changes start with an issue (some are even bigger, like a separate project, separate fork, separate repo, separate working group etc.)? e.g. see https://github.com/nodejs/node/pulls?q=is%3Apr+is%3Amerged+sort%3Acomments-desc+ (I think even the ones that don't reference an issue, it's mostly that the author forgot to add the links, but there was almost always one/multiple issues leading up to it or offline discussions/design docs). For experienced contributors this isn't always necessary e.g. there might've already been offline discussions that don't have an issue, or they already know "the temperature in the room for this", but for first-time contributors I think it's useful to mention that if you don't do this, there's a chance that a complex PR would be dragged into endless "do we even want to have it/do we even want it to work this way" discussions as listed above instead of actual code reviews.

> (FWIW for my first contribution, I started with an issue (https://github.com/nodejs/node/issues/19443) and that was a complete waste of my time)

I don't think that falls into the scope of:

> but your proposed change is complex, introduces a new feature, breaks backward compatibility, or needs a significant refactor,

It's a strightforward semver-patch PR?

Author: Ping @aduh95 WDYT? I think as more and more AI-generated PR that never get any review come in this would be more useful - for anything substantial it's better to discuss in an issue about "do we even want it" or "how the API should look like" before dumping a big PR that no one will ever review because of lack of context.

Author: Updated to add a note emphasizing

> Straightforward or non-controversial changes, such as an obvious bug fix or a small refactor, don't
> usually benefit from discussions in a separate issue, and can be sent directly as a pull request.

blocks of code, logs, or text generated by tools. If necessary, use `<details>` tags to fold
them into a collapsible section to make the discussion more readable.

## What NOT to do for your first contributions

**Refrain from automating your first contributions**. In this project, automation must
be authorized and tracked explicitly. If you have an idea for automation, please open an
issue to discuss it with other contributors first.

**Refrain from submitting large pull requests for your first contributions**. Node.js makes
technical decisions through a
[consensus-seeking process among its collaborators](./collaborator-guide.md#consensus-seeking),
with additional community feedback taken into account. Pushing a complex change through
this process can be difficult even for experienced contributors, and discouraging for
first-time contributors. It's best to start with smaller, focused pull requests that
address known issues, allowing you to familiarize yourself with the collaboration
process first. When you have to submit a large pull request after gaining some experience,
make sure to follow the [large pull request guidelines](./large-pull-requests.md).

> 

Reviewer: ```suggestion

**Refrain from opening new Pull Requests before the first one has been approved**.
This gives others the first impression that you are flooding the project.
It's OK to not get everything right on the first try, but if you open PRs faster
than they are reviewed, you are wasting an opportunity to improve, and can overwhelm
both yourself and your reviewers with notifications.
```

be authorized and tracked explicitly. If you have an idea for automation, please open an
issue to discuss it with other contributors first.

**Refrain from submitting large pull requests for your first contributions**. Node.js makes
technical decisions through a
[consensus-seeking process among its collaborators](./collaborator-guide.md#consensus-seeking),
with additional community feedback taken into account. Pushing a complex change through
this process can be difficult even for experienced contributors, and discouraging for
first-time contributors. It's best to start with smaller, focused pull requests that
address known issues, allowing you to familiarize yourself with the collaboration
process first. When you have to submit a large pull request after gaining some experience,
make sure to follow the [large pull request guidelines](./large-pull-requests.md).

**Refrain from opening new Pull Requests before the first one has been approved**.
This gives others the first impression that you are flooding the project.
It's OK to not get everything right on the first try, but if you open PRs faster
than they are reviewed, you are wasting an opportunity to improve, and can overwhelm

> both yourself and your reviewers with notifications.

Author: ```suggestion
both yourself and your reviewers with notifications.

```

### Q: There are some review comments on my pull request. How should I update my pull request to address them?

If the review comments are left with GitHub's suggestion feature, you can apply the suggestions
directly in the GitHub UI and commit the changes to your branch. Otherwise, you can push new commits to
the pull request branch. See [the pull request guide](./pull-requests.md#step-9-discuss-and-update) for
the technical how-tos.

After you push new commits, even if the previous commits have been approved, you will still need a
new approval for the last commits.

Commits in the same pull requests are usually squashed into one commit before landing, unless
there's a good justification to keep them separate - in that case, coordinate with your reviewers
about this. The commits that need to be kept separate must each pass all tests on their own and
follow the [commit message guidelines](./pull-requests.md#commit-message-guidelines).

Commit squashing is usually handled by the automation. Commits that are intended to be squashed don't
have to follow the commit message guidelines, so you don't have to constantly squash them which would

> require new approvals/CI runs.

Reviewer 2: The rationale here sounds a bit misleading: constantly squashing by itself (assuming it's done either right after pushing new commit(s) or *instead* of pushing new commit(s), which is usually the case) doesn't affect requirement for new approvals/CI, it's required for any subsequent commits anyway.

I'm not sure how to word this without sounding too distrustful to first-time contributors but the biggest issue in this context is that it adds extra burden on reviews, i.e. the changes must be doublechecked every time to see if no malicious (or sincerely unintended) code got injected into one of 'apply minor changes & rebase in main' squashes.
Aside from that, generic reasoning may apply: easier to review, easier to undo, etc. etc.

Author: The case I had in mind was different - sometimes people thought that they had to squash it themselves, so they push new commits, and keep them separate for the ease of review, and after approval they still squash it for "finalization" thinking that it matters for the commit message guidelines (which is uncessary).

Reviewer 2: Makes sense, although `constantly` feels contradictory to this case. :smile:
And it still IMHO sounds misleading, i.e. implying that non-squash push might *not* require new approval&CI.

How about adding one more short subsection or Q, explaining the 'finalization' and 'what not to do' in particular? That landing doesn't require squashing (aside from `commit-rebase` edge cases), adding commit message trailers, signing CLA, etc.
It would also be a good place to explain https://github.com/nodejs/node/labels/author%20ready label and that it neither awaits or restricts actions from author themselves.

Author: This doc is already quite big in this PR - would you mind adding that section in a follow-up PR?

Author: Updated the wording a bit to clarify how squashing is done.

> Commit squashing is usually handled by the automation. Commits that are intended to be squashed away don't
> have to follow the commit message guidelines - only the ones whose commit message will be kept after
> squashing do. If your stack of commits are already approved and the CI has passed on them, don't worry
> about squashing them yourself - the automation will take care of it during landing.

will verify that the pull request meets all the requirements before merging it, but the
triggering of the commit queue still requires human judgement and depends on volunteers.
You can ask for help either in the pull request or in the slack channel similar
to how you ask for help to get reviews and trigger CI runs.

### Q: There's a comment saying that my pull request is landed, but instead I see my pull request closed, what happened?

In this project, pull requests are typically merged by the
[commit queue](../../.github/workflows/commit-queue.yml) to amend the commit with
necessary metadata, not through the GitHub UI. Sometimes the automation or the collaborator
performing the merge can update the branches in a way that GitHub recognizes it as a merge
and shows a purple "merged" status, sometimes it's not the case.
As long as you see a comment saying that your pull request is landed, and the commits do appear
in the target branch of your pull request, don't worry, it means the change has been effectively
merged, regardless of what GitHub shows.

### Q: When will my change be released?

Commits merged into the `main` are released according to the release schedule,

> which are displayed in the [Release Working Group's pinned issues](https://github.com/nodejs/Release/issues).

Reviewer 3: ```suggestion
Commits merged into the `main` branch are released according to the release schedule of each release line.
These are displayed in the [Release Working Group's pinned issues](https://github.com/nodejs/Release/issues).
```

> 

Reviewer 3: Perhaps there should also be some description here about backporting? Commits are released first into the Current release line branch, then after a period of stabilization, they get backported to LTS release lines. Something like that ...

There is a built-in delay to releasing changes in LTS branches due to this.

Without some more explanation here, I don't think that this paragraph is very useful to new contributors. It's more likely to be confusing.

Author: Thanks, this is a FAQ I get a lot too, added a few sentences about this.

### Q: There are some review comments on my pull request. How should I update my pull request to address them?

If the review comments are left with GitHub's suggestion feature, you can apply the suggestions
directly in the GitHub UI and commit the changes to your branch. Otherwise, you can push new commits to
the pull request branch. See [the pull request guide](./pull-requests.md#step-9-discuss-and-update) for
the technical how-tos.

After you push new commits, even if the previous commits have been approved, you will still need a
new approval for the last commits.

Commits in the same pull requests are usually squashed into one commit before landing, unless
there's a good justification to keep them separate - in that case, coordinate with your reviewers
about this. The commits that need to be kept separate must each pass all tests on their own and
follow the [commit message guidelines](./pull-requests.md#commit-message-guidelines).

Commit squashing is usually handled by the automation. Commits that are intended to be squashed away don't
have to follow the commit message guidelines - only the ones whose commit message will be kept after

> squashing do. If your stack of commits are already approved and the CI has passed on them, don't worry

Reviewer 3: ```suggestion
squashing do. If your commits are already approved and the CI has passed on them, don't worry
```

otherwise it would need to be "stack of commit is ..." to be grammatically correct, but that sounds a bit strange

are showing up in other pull requests - if they do, it's very likely that they are unrelated to your
change. In this case, you can ignore those failures and ask for help to trigger the CI runs again -
often, these failures are intermittent and will disappear in the next run.

Tip: the flakes have been plaguing the project for a long time because there aren't enough volunteers
to fix them. If you have the time and interest, try volunteering to fix the ones that you are
confident with! See the
[reliability repository README](https://github.com/nodejs/reliability#protocols-in-improving-ci-reliability)
for more details on how to get started.

### Q: My pull request has enough approvals and passed CI, but it still hasn't been merged. What should I do?

Pull requests need to be merged by collaborators or triagers (normally by applying the `commit-queue`
label). The [commit queue automation](../../.github/workflows/commit-queue.yml)
will verify that the pull request meets all the requirements before merging it, but the
triggering of the commit queue still requires human judgement and depends on volunteers.

> You can ask for help either in the pull request or in the slack channel similar

Reviewer 3: ```suggestion
You can ask for help either in the pull request or in the Slack channel similar
```

branding

[commit queue](../../.github/workflows/commit-queue.yml) to amend the commit with
necessary metadata, not through the GitHub UI. Sometimes the automation or the collaborator
performing the merge can update the branches in a way that GitHub recognizes it as a merge
and shows a purple "merged" status, sometimes it's not the case.
As long as you see a comment saying that your pull request is landed, and the commits do appear
in the target branch of your pull request, don't worry, it means the change has been effectively
merged, regardless of what GitHub shows.

### Q: When will my change be released?

If the pull request is not `semver-major`, commits merged into the `main` branch are typically
first released in next "Current" release. After spending at least two weeks in the "Current" release,
a change that is not `semver-major` will be eligible to be included in the next "LTS" releases.
The "Current" and "LTS" releases follow schedules that are tracked in the
[Release Working Group's pinned issues](https://github.com/nodejs/Release/issues).

If the change is labeled `dont-land-*`, it will be held back from releases in the specified branch until
the label is removed.

> If the change is labeled `semver-major`, it will only not be included in the next "Current" release.

Reviewer 3: ```suggestion
If the change is labeled `semver-major`, it will not be included in the next "Current" release.
```
